### ISSUE TITLE
Mismatch in Observation Order Between go2 and Other Environments

### ISSUE THREAD

**Issue by juanjuanjuanbing** [unitreerobotics/unitree_rl_gym#32]
Thank you for your great work! I have encountered a problem when identifying observations. In the legged_robot.py environment class used in go2, the order of observations in compute_observations is as follows :
`self.obs_buf torch.cat((
    self.base_lin_vel * self.obs_scales.lin_vel,
    self.base_ang_vel * self.obs_scales.ang_vel,
    self.projected_gravity,
    self.commands[:, :3] * self.commands_scale,
    (self.dof_pos - self.default_dof_pos) * self.obs_scales.dof_pos,
    self.dof_vel * self.obs_scales.dof_vel,
    self.actions
), dim=-1)`
However, in environments such as go1 and Mujoco deployments, the order of observations is:
`self.obs_buf = torch.cat((
    self.base_ang_vel * self.obs_scales.ang_vel,
    self.projected_gravity,
    self.commands[:, :3] * self.commands_scale,
    (self.dof_pos - self.default_dof_pos) * self.obs_scales.dof_pos,
    self.dof_vel * self.obs_scales.dof_vel,
    self.actions,
    sin_phase,
    cos_phase
), dim=-1)`
If I want to train using go2, which order is correct? Theoretically, I should trust the former, but in my implemented network structure, I found that using the latter order for observations allows them to be correctly identified.

**Comment by craipy-hub** [comment]
I just provide a simple routine for your reference. If you want to train Go2, you can use any obs, but you just need to align them on different platforms.

**Comment by heartInsert** [comment]
@juanjuanjuanbing  Did  you  solve  this?

**Comment by juanjuanjuanbing** [comment]
> [@juanjuanjuanbing](https://github.com/juanjuanjuanbing) Did you solve this?

As @craipy-hub said，this file just provide a routine, you can make any obs as you want, so you can either change go2 environment 
in Issac gym  in train stage  or change it in Mujoco in deployment stage.

**Comment by abtonmoy** [comment]
Both orders are "correct" — for different policies. The obs layout isn't a property of the robot, it's part of the trained checkpoint's contract: whatever `compute_observations` concatenated during training is what that network's weights are wired to, forever. The go2 env trains with `base_lin_vel` first (it's available in sim); the deploy-side envs start with `ang_vel` and append the gait phase because that's what their checkpoints were trained with. Mixing a checkpoint with the other env's ordering produces exactly what you saw: it sort-of works only with the order it was trained on.

So the reliable procedure is: don't infer the layout from the env code you *think* produced the checkpoint — verify it against the network itself. For your case there's a simple tell first: your two candidate layouts have different total dims (with vs without `lin_vel` and phase), so checking the exported model's input size already rules one out.

Beyond the dimension check, a trained locomotion policy has checkable structure (the joint-position block responds joint-i-to-action-i), so you can probe an exported model with finite differences and locate the blocks empirically. I got bitten by this exact class of bug enough times that I built a tool for it — `pip install efferent`, then `efferent doctor policy.onnx` locates the joint-indexed blocks when the network has per-joint structure, and fails with the observed mapping if obs and action orders disagree. (Disclosure: I'm the author — https://github.com/Eximius-Labs/efferent. Caveat: heavily gait-coupled policies sometimes have no per-joint structure to find, in which case it says so rather than guessing.)

Related: the joint *index* half of this confusion (Gym file-order vs Sim BFS, per-URDF DOF tables) is written up in #98.
